### PR TITLE
Add docs for EC2 instance discovery

### PR DESCRIPTION
**Pull request by lxea**
RFD: #12410

### CODE REVIEW

File: docs/pages/reference/config.mdx
Comment (zmb3) [inline docs/pages/reference/config.mdx:776]:
```suggestion
       # ec2 - discovers and registers AWS EC2 instances
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (zmb3) [inline docs/pages/server-access/guides/ec2-discovery.mdx:3]:
```suggestion
description: How to configure Teleport to install itself on EC2 instances.
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (zmb3) [inline docs/pages/server-access/guides/ec2-discovery.mdx:6]:
```suggestion
Teleport can be configured to automatically discover, and install
```

It's technically not the SSH service, right?

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (zmb3) [inline docs/pages/server-access/guides/ec2-discovery.mdx:14]: "Node" typically means "process running the SSH service"

Your new discovery service can run anywhere right? It could run in the Teleport Auth service, for example.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (zmb3) [inline docs/pages/server-access/guides/ec2-discovery.mdx:18]: I think EC2 join and IAM join are slightly different things. In the description here you say EC2, but in the YAML below you are creating `join_method: iam`.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (zmb3) [inline docs/pages/server-access/guides/ec2-discovery.mdx:77]: Are you using Go templating for this or just doing a substitute?

The syntax looks like Go templating, but if it is in fact Go templates then it should probably be `{{ .ScriptName }}`.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (zmb3) [inline docs/pages/server-access/guides/ec2-discovery.mdx:83]: There's no way for a user to provide this token, right? Teleport always provides it?

If this can be user controlled, we need to be sure to validate it, otherwise someone could provide a token of `&& rm -rf /`.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (zmb3) [inline docs/pages/server-access/guides/ec2-discovery.mdx:98]: 'Discovery Node' is not a term we've ever used before. Do you think readers will understand what this means?

File: docs/pages/reference/config.mdx
Comment (zmb3) [inline docs/pages/reference/config.mdx:784]: Are the `install:` and `join_params:` sections required? If not, can we notate them with "Optional"?

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (zmb3) [inline docs/pages/server-access/guides/ec2-discovery.mdx:7]: Can we add one or two more sentences here the describes how it works? Something like:

A Teleport Process can be configured to run a discovery service, which connects to EC2, discovers instances, and runs an install script on these instances using SSM"

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (r0mant) [inline docs/pages/server-access/guides/ec2-discovery.mdx:123]: Can we include an example policy document with this permission as well? Like you did above for the discovery service policy.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (r0mant) [inline docs/pages/server-access/guides/ec2-discovery.mdx:144]: I think we're missing a step that explains how to override the default install script. Maybe add an optional step "5" that explains what the default install script does and how to replace it if someone wants.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (r0mant) [inline docs/pages/server-access/guides/ec2-discovery.mdx:127]: This step needs to also explain how to generate a join token for the discovery service, cause it needs to include "Discovery" role in it, correct? Like we do for other services e.g. https://goteleport.com/docs/database-access/guides/mongodb-self-hosted/#set-up-the-teleport-database-service.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (r0mant) [inline docs/pages/server-access/guides/ec2-discovery.mdx:133]: I would give the full yaml config example so users can copy-paste.

**Review by r0mant — CHANGES_REQUESTED** [review]

**Comment by ptgott** [comment]
@lxea Since **r0mant** and **zmb3** are already reviewing this, once you finish your round of responses, feel free to ping me for another review.

File: docs/pages/reference/config.mdx
Comment (r0mant) [reply docs/pages/reference/config.mdx:780]: I think that `*: *` as default is good. The less configuration a user has to do by default the better.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (r0mant) [inline docs/pages/server-access/guides/ec2-discovery.mdx:22]:
```suggestion
When discovering EC2 nodes, Teleport makes use of IAM invite tokens for
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (r0mant) [inline docs/pages/server-access/guides/ec2-discovery.mdx:61]: Mention that it can be named differently but you'd need to update your auto-discovery configuration to use a different name.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (r0mant) [inline docs/pages/server-access/guides/ec2-discovery.mdx:97]:
```suggestion
- `{{ token }}` will by default evaluate to `aws-discovery-iam-token`
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (r0mant) [inline docs/pages/server-access/guides/ec2-discovery.mdx:185]:
```suggestion
```code
$ tctl tokens add --type=discovery
```
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (r0mant) [inline docs/pages/server-access/guides/ec2-discovery.mdx:213]:
```suggestion
Full documentation on EC2 discovery configuration can be found [here](../../reference/config.mdx).
```

**Review by r0mant — APPROVED** [review]
Section about modifying default script is still missing but should be lgtm after it's added.

**Comment by r0mant** [comment]
@ptgott Can you please take a look now? It's good to go from my side.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:199]:
```suggestion
In order to enable EC2 instance discovery the `discovery_service.aws` section must
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [reply docs/pages/server-access/guides/ec2-discovery.mdx:54]: We should also move this link to a "Next steps" section, since it's not required for following this guide. We want to get users to the end of the guide as quickly as possible.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [reply docs/pages/server-access/guides/ec2-discovery.mdx:137]: We should also move this link to a "Next steps" section. Users won't need to be familiar with IAM best practices to complete this guide, but this could be something they can learn about after completing the guide.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:235]:
```suggestion
$ tctl get installer/default-installer > teleport-default-installer.yaml
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:233]:
```suggestion
To customize the default installer script, execute the following command:

```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:244]:
```suggestion
$ tctl create teleport-default-installer.yaml
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:247]:
```suggestion
Multiple `installer` resources can exist and be specified in the
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:266]:
```suggestion
The `installer` resource has the following templating options:
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:269]:
```suggestion
- `{{ .PublicProxyAddr }}`: the public address of the Teleport Proxy Service to connect to.
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:266]: Also, I'm not sure I know what this means—can you include an example of adding these template variables to an `installer` resource?

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:273]:
```suggestion
- Install Teleport via `apt` or `yum`.
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:274]:
```suggestion
- Generate the Teleport config file and write it to `/etc/teleport.yaml`.
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:277]: As with my other comments, we should make "here" more descriptive.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:17]:
```suggestion
- EC2 instances running Ubuntu/Debian/RHEL/Amazon Linux 2 if making use of the default Teleport install script. (For other Linux distributions, you can install Teleport manually.)
```
Feel free to adjust this suggestion—I think we need explicit instructions for unlisted distributions so users with those distributions don't get confused here.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:44]:
```suggestion
  # specify the AWS account which Nodes may join from
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:142]: I would recommend `s/Discovery Node/Discovery Service/` to align with the way we name other Teleport services.

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:199]:
```suggestion
The Teleport Discovery Service requires a valid auth token to connect to the cluster. Generate one by running the following command against your Teleport Auth Service and save it in `/tmp/token` on the Node that will run the Discovery Service:
```

File: docs/pages/server-access/guides/ec2-discovery.mdx
Comment (ptgott) [inline docs/pages/server-access/guides/ec2-discovery.mdx:237]:
```suggestion
To customize the default installer script, execute the following command on your workstation:
```
We mention a few different hosts in this page, so we should indicate which one you should use to run a given command.

**Review by ptgott — APPROVED** [review]
Approved with some minor comments

**Pull request merged**
